add app management section (#1704)
giantswarm/docs: src/content/advanced/gitops/apps/add_app_template.md

---
linkTitle: Add a new App Template
title: Add a new App Template
description: Learn how to create application templates to deploy apps using GitOps.
weight: 50
menu:
  main:
    identifier: advanced-gitops-apps-add-app-template
    parent: advanced-gitops-apps
last_review_date: 2023-02-10
user_questions:
  - How can I create an template for app deployment in GitOps?
owner:
  - https://github.com/orgs/giantswarm/teams/team-honeybadger
---

This document is part of the documentation to use GitOps with Giant Swarm App Platform. You can find more information about the [App Platform in our docs](/platform-overview/app-platform/).

> # Add a new App Template to the repository

Reviewer: On the  capitalisation of "App Template" and others:

Throughout this file, the term "App Template" is capitalised - this makes me think that there is a type called `AppTemplate` especially when read alongside `Config Map` or `App Custom Resource`.

As `App Template` is not a type, but a pattern used to define a type, I would consider not capitalising this to avoid confusion to the reader.

Author: changed everywhere but the title (title case)

> In order to avoid adding the same application from scratch across all your clusters, you can prepare App Templates providing a pre-configured version of an App. This also allows you to manage and version an app's configuration even if the app itself is not yet installed in any cluster.

Reviewer: ```suggestion
To avoid duplication caused by adding the same application from scratch across all your clusters, you can prepare App Templates providing a pre-configured version of an App. This also allows you to manage and version an app's configuration even if the app itself is not yet installed in any cluster.
```

## Example

An example of an App Template is available in the [gitops-template repository "bases/apps/nginx-ingress-controller"](https://github.com/giantswarm/gitops-template/tree/main/bases/apps/nginx-ingress-controller).

## Export environment variables

> __Note__: Management Cluster codename, Organization name, Workload Cluster name and several App-related values are needed in multiple places across this instructions, the least error prone way of providing them is by exporting them as environment variables:

Reviewer: ```suggestion
__Note__: Management Cluster codename, Organization name, Workload Cluster name and several App-related values are needed in multiple places across these instructions. The least error prone way of providing them is by exporting them as environment variables:
```

```nohighlight
export WC_NAME=CLUSTER_NAME
export APP_NAME="${WC_NAME}-APP_NAME"
export APP_VERSION=APP_VERSION
export APP_CATALOG=APP_CATALOG
export APP_NAMESPACE=APP_NAMESPACE
# OPTIONAL more info about App Configuration can be found at https://docs.giantswarm.io/getting-started/app-platform/app-configuration/#basic-values-merging-example
export APP_USER_VALUES=CONFIGMAP_OR_SECRET_PATH
```

## Setting up directory tree structure for managing apps

> 1. Go to the `apps` directory and prepare a directory for the new App Template:

Reviewer: ```suggestion
1. Go to the `bases/apps` directory and create a new directory for the new App Template:
```

    ```nohighlight
    cd bases/apps/
    mkdir ${APP_NAME}
    ```

> 1. Go to the newly created directory and use [the kubectl-gs plugin](https://github.com/giantswarm/kubectl-gs) to generate the [App CR](https://docs.giantswarm.io/ui-api/kubectl-gs/template-app/):

Reviewer: ```suggestion
1. Navigate to the newly created directory and use [the kubectl-gs plugin](https://github.com/giantswarm/kubectl-gs) to generate the [App CR](https://docs.giantswarm.io/ui-api/kubectl-gs/template-app/):
```

    ```nohighlight
    cd ${APP_NAME}/
    kubectl gs template app \
    --app-name ${APP_NAME} \
    --catalog ${APP_CATALOG} \
    --cluster-name ${WC_NAME} \
    --name ${APP_NAME} \
    --target-namespace ${APP_NAMESPACE} \
    --version ${APP_VERSION} > appcr.yaml
    ```

> __Note__: you most likely want to provide a default configuration, so add any of the below flags to the previous command:

Reviewer: ```suggestion
__Note__: you most likely want to provide a default configuration, and optionally additional secrets for your application. The flags below can be used to achieve this by adding them to the previous command:
```

    ```nohighlight
    --user-configmap ${APP_USER_VALUES}
    --user-secret ${APP_USER_VALUES}
    ```

__Note__: We're including `${cluster_name}` in the app name to avoid a problem when two or more clusters in the same organization want to deploy the same app with its default name.

Reference [the App Configuration](https://docs.giantswarm.io/app-platform/app-configuration/) for more details about how to properly create the respective ConfigMaps or Secrets.

> In case you used `kubectl gs` command you realized the output is an App Custom Resource plus the Config Map. In case you want to manage the values in plain YAML, you could rely on the ConfigMap generator feature of [Kustomize](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#generating-resources).

Reviewer: ```suggestion
In case you used `kubectl gs` command you may have realized the output is an App Custom Resource plus the Config Map. In case you want to manage the values in plain YAML, you could rely on the ConfigMap generator feature of [Kustomize](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#generating-resources).
```

> In case you used `kubectl gs` command you realized the output is an App Custom Resource plus the Config Map. In case you want to manage the values in plain YAML, you could rely on the ConfigMap generator feature of [Kustomize](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#generating-resources).

Reviewer: Use of both `Config Map` and `ConfigMap` here. As `ConfigMap` is a type, I would not separate the words.

__Warning__: It can not be used for the Secrets as they need to be encrypted before commit into the Git Repository. Refer to our [adding an App](./add_appcr.md) docs to check how to add and encrypt a Secret.

> 1. Now it is time to create the `kustomization.yaml` file, adding the optional Secret/Config Map as resources and/or using a ConfigMapGenerator to manage plain configuration:

Reviewer: Similarly to above, `ConfigMap` is the type.

---
linkTitle: Add a new App Template
title: Add a new App Template
description: Learn how to create application templates to deploy apps using GitOps.
weight: 50
menu:
  main:
    identifier: advanced-gitops-apps-add-app-template
    parent: advanced-gitops-apps
last_review_date: 2023-02-10
user_questions:
  - How can I create an template for app deployment in GitOps?
owner:
  - https://github.com/orgs/giantswarm/teams/team-honeybadger
---

> 

Reviewer 2: I was thinking if we can add an intro in each of these 'Child' entries so in case someone comes from Google she knows that is part of a whole thing. WDYT?

Author: added a small paragraph with a link to platform overview

# Add a new App Template to the repository

> In order to avoid adding the same application from scratch across all your clusters, you can prepare App Templates that provide a pre-configured version of an App. This also allows you to manage and version an app's config even if the app itself is not yet installed in any cluster.

Reviewer 2: ```suggestion
In order to avoid adding the same application from scratch across all your clusters, you can prepare App Templates providing a pre-configured version of an App. This also allows you to manage and version an app's configuration even if the app itself is not yet installed in any cluster.
```

## Example

An example of an App Template is available in the [gitops-template repository "bases/apps/nginx-ingress-controller"](https://github.com/giantswarm/gitops-template/tree/main/bases/apps/nginx-ingress-controller).

## Export environment variables

> **Note**, Management Cluster codename, Organization name, Workload Cluster name and several App-related values are needed in multiple places across this instruction, the least error prone way of providing them is by exporting as environment variables:

Reviewer 2: ```suggestion
**__Note__**, Management Cluster codename, Organization name, Workload Cluster name and several App-related values are needed in multiple places across this instruction, the least error prone way of providing them is by exporting them as environment variables:
```

> **__Note__**, Management Cluster codename, Organization name, Workload Cluster name and several App-related values are needed in multiple places across this instruction, the least error prone way of providing them is by exporting them as environment variables:

Reviewer 2: ```suggestion
**__Note__**, Management Cluster codename, Organization name, Workload Cluster name and several App-related values are needed in multiple places across this instructions, the least error prone way of providing them is by exporting them as environment variables:
```

> **__Note__**, Management Cluster codename, Organization name, Workload Cluster name and several App-related values are needed in multiple places across this instructions, the least error prone way of providing them is by exporting them as environment variables:

Reviewer 2: ```suggestion
__Note__: Management Cluster codename, Organization name, Workload Cluster name and several App-related values are needed in multiple places across this instructions, the least error prone way of providing them is by exporting them as environment variables:
```

Reviewer 2: I have checked how it is done in other pages

```nohighlight
export WC_NAME=CLUSTER_NAME

> export APP_NAME="\${cluster_name}-APP_NAME"

Reviewer 2: ```suggestion
export APP_NAME=$WC_NAME"-APP_NAME"
```
we use vars but it is not automatic replaces so it gets confusion IMO

export APP_VERSION=APP_VERSION
export APP_CATALOG=APP_CATALOG
export APP_NAMESPACE=APP_NAMESPACE

> # OPTIONAL

Reviewer 2: ```suggestion
# OPTIONAL more info about App Configuration can be found at https://docs.giantswarm.io/getting-started/app-platform/app-configuration/#basic-values-merging-example
```

export APP_USER_VALUES=CONFIGMAP_OR_SECRET_PATH
```

## Setting up directory tree structure for managing apps

1. Go to the `apps` directory and prepare a directory for the new App Template:

    ```nohighlight
    cd bases/apps/
    mkdir ${APP_NAME}
    ```

1. Go to the newly created directory and use [the kubectl-gs plugin](https://github.com/giantswarm/kubectl-gs) to generate the [App CR](https://docs.giantswarm.io/ui-api/kubectl-gs/template-app/):

    ```nohighlight
    cd ${APP_NAME}/

>     kubectl gs template app \
>     --app-name ${APP_NAME} \
>     --catalog ${APP_CATALOG} \
>     --cluster ${WC_NAME} \
>     --name ${APP_NAME} \
>     --namespace ${APP_NAMESPACE} \
>     --version ${APP_VERSION} > appcr.yaml

Reviewer 2: ```suggestion
    kubectl gs template app \
    --app-name ${APP_NAME} \
    --catalog ${APP_CATALOG} \
    --cluster-name ${WC_NAME} \
    --name ${APP_NAME} \
    --target-namespace ${APP_NAMESPACE} \
    --version ${APP_VERSION} > appcr.yaml
```

```
Flag --cluster has been deprecated, use --cluster-name instead
Flag --namespace has been deprecated, use --target-namespace instead```

    ```

>     **Note**, you most likely want to provide a default configuration, so add any of the below flags to the previous command:

Reviewer 2: ```suggestion
**__Note__**, you most likely want to provide a default configuration, so add any of the below flags to the previous command:
```
For convention

> **__Note__**, you most likely want to provide a default configuration, so add any of the below flags to the previous command:

Reviewer 2: ```suggestion
__Note__: you most likely want to provide a default configuration, so add any of the below flags to the previous command:
```

    ```nohighlight
    --user-configmap ${APP_USER_VALUES}
    --user-secret ${APP_USER_VALUES}
    ```

>     **Note**, We're including `${cluster_name}` in the app name to avoid a problem when two or more clusters in the same organization want to deploy the same app with its default name.

Reviewer 2: ```suggestion
**__Note__**, We're including `${cluster_name}` in the app name to avoid a problem when two or more clusters in the same organization want to deploy the same app with its default name.
```

> **__Note__**, We're including `${cluster_name}` in the app name to avoid a problem when two or more clusters in the same organization want to deploy the same app with its default name.

Reviewer 2: ```suggestion
__Note__: We're including `${cluster_name}` in the app name to avoid a problem when two or more clusters in the same organization want to deploy the same app with its default name.
```

>     Reference [the App Configuration](https://docs.giantswarm.io/app-platform/app-configuration/) for more details about

Reviewer 2: I am confused with indentation. We don't use indent in any paragraph usually. Example -> https://raw.githubusercontent.com/giantswarm/docs/main/src/content/getting-started/app-platform/app-bundle/index.md

    how to properly create the respective ConfigMaps or Secrets.

>     If you want to provide a default config, we can use use the ConfigMap generator feature of `kustomize`. This allows us to use a pure YAML file for the configuration, without wrapping it into a ConfigMap. Still, for Secrets we need to encrypt them as a Secret object and the generator approach won't work. Refer to our [adding an App](./add_appcr.md) docs to check how to add and encrypt a Secret. For configuration that can be used as a ConfigMap, just add the content to a `default_config.yaml` file.

Reviewer 2: ```suggestion
In case you used `kubectl gs` command you realized the output is an App Custom Resource plus the Config Map. In case you want to manage the values in plain YAML, you could rely on the ConfigMap generator feature of [Kustomize](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#generating-resources).
```

> 1. Add the `kustomization.yaml` file, adding an optional Secret as a resource and a ConfigMapGenerator for plain text config:

Reviewer 2: ```suggestion
1. Now it is time to create the `kustomization.yaml` file, adding the optional Secret/Config Map as resources and/or using a ConfigMapGenerator to manage plain configuration:
```

    ```yaml
    apiVersion: kustomize.config.k8s.io/v1beta1
    kind: Kustomization
    buildMetadata: [originAnnotations]
    # default config block start - include if you provide default config
    configMapGenerator:
      - files:
        - values=default_config.yaml
        name: ${cluster_name}-nginx-ingress-controller-values
    generatorOptions:
      disableNameSuffixHash: true
    # default config block end
    resources:
      - appcr.yaml

>       - secret.enc.yaml # only if you provide default Secret

Reviewer 2: ```suggestion
      - secret.enc.yaml # only if you provide the default Secret
      # You can add here the config map in case of generate it via kubectl gs command or manually
```

    ```

> At this point, you should have a ready App Template.

Reviewer 2: can we reference to the doc entry where the app template is used?

Author: added more explicitly, it was implicitly in the next steps.
